modules/netweaver/main.tf

```diff
@@ -95,38 +90,51 @@ resource "google_compute_instance" "gcp_nw" {
   }
 
   boot_disk {
-    auto_delete = "${var.autodelete_disk}"
+    auto_delete = var.autodelete_disk
 
     device_name = "${var.instance_name}-${var.device_0}"
 
     initialize_params {
       image = "projects/${var.linux_image_project}/global/images/family/${var.linux_image_family}"
-      size  = "${var.boot_disk_size}"
-      type  = "${var.boot_disk_type}"
+      size  = var.boot_disk_size
+      type  = var.boot_disk_type
     }
   }
 
   network_interface {
-    subnetwork         = "${var.subnetwork}"
-    subnetwork_project = "${var.project_id}"
-    access_config      = "${local.access_config[var.public_ip]}"
+    subnetwork         = var.subnetwork
+    subnetwork_project = var.project_id
+
+    dynamic "access_config" {
+      for_each = var.public_ip == 1 ? ["external_ip"] : []
+      content {
+        # assign an ephemeral external IP address
+        nat_ip = null
+      }
+    }
+
+  }
+
+  metadata = {
+    instanceName           = var.instance_name
+    instanceType           = var.instance_type
+    post_deployment_script = var.post_deployment_script
+    subnetwork             = var.subnetwork
+    usrsapSize             = var.usr_sap_size
+    sapmntSize             = var.sap_mnt_size
+    swapSize               = var.swap_size
+    sap_deployment_debug   = var.sap_deployment_debug
+    startup-script         = var.startup_script
+    publicIP               = var.public_ip
   }
 
-  metadata {
-    instanceName           = "${var.instance_name}"
-    instanceType           = "${var.instance_type}"
-    post_deployment_script = "${var.post_deployment_script}"
-    subnetwork             = "${var.subnetwork}"
-    usrsapSize             = "${var.usr_sap_size}"
-    sapmntSize             = "${var.sap_mnt_size}"
-    swapSize               = "${var.swap_size}"
-    sap_deployment_debug   = "${var.sap_deployment_debug}"
-    startup-script         = "${var.startup_script}"
-    publicIP               = "${var.public_ip}"
+  lifecycle {
+    # Ignore changes in the instance metadata, since it is modified by the SAP startup script.
+    ignore_changes = [metadata]
```

Comment:
I looked into that and the only way to make it work was to ignore the full metadata. I think there was an issue in particular with the startup-script metadata, which needs to be removed in the second pass. So the only possible way was to ignore the full metadata.

Reply:
Coming back to this, as explained in the [terraform documentation](https://www.terraform.io/docs/configuration/resources.html#lifecycle-lifecycle-customizations):

> You can also ignore specific map elements by writing references like tags["Name"] in the ignore_changes list, though with an important caveat: the ignoring applies only to in-place updates to an existing key. Adding or removing a key is treated by Terraform as a change to the containing map itself rather than to the individual key

To verify this, I tried ignoring only the keys that are being deleted plus the one that is being added (status):
```
    ignore_changes = [
      metadata["sap_hana_sidadm_password"],
      metadata["sap_hana_system_password"],
      metadata["startup-script"],
      metadata["status"],
    ]
```
Once the instance has been initialized, I still get this diff:
```
      ~ metadata             = {
...
            "sap_hana_sid"               = "D10"
          + "sap_hana_sidadm_password"   = "myPassword123"
            "sap_hana_sidadm_uid"        = "900"
          + "sap_hana_system_password"   = "myPassword123"
          + "startup-script"             = <<~EOT
                #!/bin/bash
```
Since the SAP startup script removes those three keys, the change is considered as a change to the dictionary and not the missing keys. This confirms that we need to ignore changes to the whole metadata if we want to avoid those diffs when running terraform plan.

Reply:
Got it, thanks for investigating!

FWIW, in a future update it might be worth changing the script to empty/null these values instead of removing them entirely. It would be good to keep the `ignore_changes` radius as small as possible.